### PR TITLE
Fixes and tests for RTP8i, RTP8f

### PR DESCRIPTION
**Pull request by psolstice**

### CODE REVIEW

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 1]
```diff
@@ -3013,4 +3013,57 @@ public void onPresenceMessage(PresenceMessage messages) {
...
+			channel.presence.enter(null, completionWaiter);
+
+			ably.connection.connect();
+
+			completionWaiter.waitFor(1);
+			assertFalse("Verify enter() failed", completionWaiter.success);
+
+			/* Now clientId is known to be "*" and subsequent enter() should fail immediately */
+			try {
+				channel.presence.enter(null, null);
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: Given how easy it is to add `update` and `leave` as well, why not add them as well to this test.

Comment (psolstice) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: This line is supposed to be the test for RTP8f. `clientId` is '*' at the moment and enter() fails as expected

Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: Yes, I know that, but see https://docs.ably.io/client-lib-development-guide/features/#RTP9e , https://docs.ably.io/client-lib-development-guide/features/#RTP10e and https://docs.ably.io/client-lib-development-guide/features/#RTP15a.  For efficiency, this test could cover this behaviour for all of those methods so why not do that now?

Comment (psolstice) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: sure, later today

File: lib/src/main/java/io/ably/lib/realtime/Presence.java [hunk 2]
```diff
@@ -425,7 +425,12 @@ public void enterClient(String clientId, Object data) throws AblyException {
 	 * @throws AblyException
 	 */
 	public void enterClient(String clientId, Object data, CompletionListener listener) throws AblyException {
-		Log.v(TAG, "enterClient(); channel = " + channel.name + "; clientId = " + clientId);
+		Log.v(TAG, "enterClient(); channel = " + channel.name + "; clientId = " + (clientId != null ? clientId : "(null)"));
+		if (clientId == null && "*".equals(channel.ably.auth.clientId)) {
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/realtime/Presence.java]: There is a static final for "*"

Comment (psolstice) [inline lib/src/main/java/io/ably/lib/realtime/Presence.java]: Yes, in Auth but it's private

Comment (psolstice) [inline lib/src/main/java/io/ably/lib/realtime/Presence.java]: nonetheless it looks better with static final, I'll make it public

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 3]
```diff
@@ -3013,4 +3013,73 @@ public void onPresenceMessage(PresenceMessage messages) {
...
+
+			/* enter without attaching first */
+			Channel channel = ably.channels.get("enter_before_clientid_is_known"+testParams.name);
+			CompletionWaiter completionWaiter = new CompletionWaiter();
+			channel.presence.enter(null, completionWaiter);
+
+			ably.connection.connect();
+
+			completionWaiter.waitFor(1);
+			assertFalse("Verify enter() failed", completionWaiter.success);
```
Comment (paddybyers) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: I think the test should get the error from the CompletionWaiter and verify that it has the expected error code

**Comment by paddybyers** [comment]
@psolstice I added https://github.com/ably/ably-java/pull/294/commits/2076a2913febe23994695f06fc3dfbdf72acdef8; pls review

**Pull request merged**